Move stonith_admin into a library

clumens opened the pull request:
At the least, I sitll need to add a bunch of documentation to the headers.  Return codes will need to be synced up with that PR.  I haven't really taken much care with return values at the moment because of that.  Here's some other things to think about:

* The pcmk__preproc_cmdline patch ends up being totally standalone, but I think it's worth it.
* I got rid of the pcmk__output_fnew/pcmk__output_dnew patch, but it could come back.  Nothing ended up needing it.
* Some of the public functions in lib/pacemaker/pcmk_stonith.c are just wrappers around the private versions.  I thought it was important to have both to keep the completeness of both APIs.
* Not everything that stonith_admin does ended up getting turned into API functions.  For instance, the register and remove device blocks in the big case statement are unchanged.  stonith_admin doesn't actually do anything around these - it just calls into `st->cmds`.  There didn't seem like enough there to warrant making a separate function, though maybe I should anyway just to keep it complete.
* Testing has been pretty minimal, but I wanted to get more people looking at the direction first.

kgaillot commented on include/pacemaker.h:
```diff
@@ -0,0 +1,137 @@
+ * \brief List all installed STONITH agents.
+ *
+ * \note If the xml parameter is not NULL, it will be freed first and the previous
+ *       contents lost.
```
FYI you can say "\p xml" instead of "the xml parameter" and doxygen will show it in fixed-width font in the documentation (though the effect is subtle). pcmk_stonith_history() needs this note as well.

kgaillot commented on include/pacemaker.h:
```diff
@@ -0,0 +1,137 @@
+                            char *target, int timeout);
+
+int pcmk_stonith_register_level(stonith_t *st, unsigned int st_opts, char *target,
+                                int fence_level, stonith_key_value_t *devices);
```
FYI, when we eventually refactor stonith_t, I plan on getting rid of stonith_key_value_t and using a list of pcmk_nvpair_t instead. We can deal with that when the refactor happens.

kgaillot commented on lib/pacemaker/pcmk_stonith.c:
```diff
@@ -0,0 +1,485 @@
+static pcmk__supported_format_t formats[] = {
+    PCMK__SUPPORTED_FORMAT_XML,
+    { NULL, NULL, NULL }
+};
```
This will likely be used for all public APIs, so maybe we can put it into a new pcmk_output.c along with prologue() and epilogue(). I'm sure we'll end up with libpacemaker messages to register, so that could go there too when the time comes.

clumens replied:
I was curious if prologue and epilogue might differ between major portions of the API.  Like, will whatever we pull out of crm_mon need something different?  At the moment I don't think so but maybe it would be instructive to lib-ify something from crm_mon and find out.

kgaillot replied:
It doesn't matter much since it's all internal, but I can't think of anything that would need to be done differently (vs additional work that could be in its own function).

kgaillot commented on lib/pacemaker/pcmk_stonith.c:
```diff
@@ -0,0 +1,485 @@
+    { NULL, NULL, NULL }
+};
+
+static GMainLoop *mainloop = NULL;
```
I hadn't thought about needing a mainloop for some requests. Do you know what happens if the caller has their own mainloop defined? Can two peacefully coexist?

clumens replied:
Yes, it is possible to have nested main loops.  That's how GTK handles stuff like dialog boxes.  There is a very, very brief mention of it in the glib main loop docs.

kgaillot commented:
@aleksei-burlakov @gao-yan @tomjelinek This is the proposed model for a high-level pacemaker C API. We're using stonith_admin as the first case since it's fairly simple. crm_mon would be next. If we one day have other language bindings, they would be wrappers around these.

Take a look at e.g. pcmk_stonith_history() and see if the design makes sense for your purposes. The idea is that each API function will take an xmlNode where it will put its formatted output (the same as if --output-as=xml were given to the equivalent command-line tool), and other arguments more or less equivalent to the equivalent tool's command-line options.

kgaillot commented:
It does occur to me that we're favoring "fencing" now over "stonith" (since the latter can imply power fencing only), so we should probably use pcmk_fence_ as the prefix.

clumens commented on include/pacemaker.h:
```diff
@@ -0,0 +1,137 @@
+ */
+int pcmk_stonith_history(xmlNodePtr *xml, stonith_t *st, unsigned int st_opts,
+                         char *target, int timeout, bool quiet, int verbose,
+                         bool broadcast, bool cleanup);
```
For the moment at least, they are all the same in everything stonith_admin does.  I've removed the parameter (which is nice - I'm always a fan of fewer arguments) and added a hidden st_opts variable to pcmk_stonith.c.

clumens commented:
> It does occur to me that we're favoring "fencing" now over "stonith" (since the latter can imply power fencing only), so we should probably use pcmk_fence_ as the prefix.

I like it.  It's a little bit shorter, too.

clumens commented on include/pacemaker.h:
```diff
@@ -0,0 +1,137 @@
+ * \param[in] timeout   How long to wait for the operation to complete (in seconds).
+ * \param[in] tolerance
+ *
+ * \return 0 on success, or various error codes on error.
```
Yeah, I'll go through and redo all the return code stuff once that PR gets merged.  The first thing I'll do is just change all the types and docs.  The second thing will be to actually think about what out of the enum everything should be returning.

clumens commented on include/pacemaker.h:
```diff
@@ -0,0 +1,137 @@
+int pcmk_stonith_last_fenced(xmlNodePtr *xml, const char *target, bool as_nodeid);
+
+int pcmk_stonith_list_targets(xmlNodePtr *xml, stonith_t *st, unsigned int st_opts,
+                              char *target, int timeout);
```
I'm a little confused by the uses of target, device, name, and agent throughout these files I've made.  This would be a nice time to straighten all that out too.

kgaillot replied:
target = node being fenced
agent = a fence agent executable
device = stonith-class resource configured in the CIB (agent + parameters)
name = arbitrary label for logging (e.g. "node1 successfully fenced for <name>")

clumens commented:
Another potential problem - if we're ever going to want to have versions of these functions that return something other than XML, we'll need to give them different names.  Does it make sense to encode the output type in the function name somehow, without making the names too unwieldy?

kgaillot commented:
> Another potential problem - if we're ever going to want to have versions of these functions that return something other than XML, we'll need to give them different names. Does it make sense to encode the output type in the function name somehow, without making the names too unwieldy?

I'm leaning against it. The chances of adding another format (vs offering an XML-to-whatever converter) are pretty low.

clumens commented on include/pacemaker.h:
```diff
@@ -0,0 +1,137 @@
+ * \param[in] action    The action to perform.
+ * \param[in] name
+ * \param[in] timeout   How long to wait for the operation to complete (in seconds).
+ * \param[in] tolerance
```
I'll mark it as seconds for now because that's what it currently does, and make a note to switch things to ms later.

clumens commented:
I've addressed several comments and updated a lot of docs.  I haven't addressed the biggest things, but I'm going to push the current state so it doesn't get lost over the break.

aleksei-burlakov commented:
> @aleksei-burlakov @gao-yan @tomjelinek This is the proposed model for a high-level pacemaker C API. We're using stonith_admin as the first case since it's fairly simple. crm_mon would be next. If we one day have other language bindings, they would be wrappers around these.
> 
> Take a look at e.g. pcmk_stonith_history() and see if the design makes sense for your purposes. The idea is that each API function will take an xmlNode where it will put its formatted output (the same as if --output-as=xml were given to the equivalent command-line tool), and other arguments more or less equivalent to the equivalent tool's command-line options.

I've looked at the `pcmk__fence_history` function. As I understand this is exactly what was previously discussed. It looks good for me, but I would also like to hear the feedback from @gao-yan.

clumens commented on include/pacemaker.h:
```diff
@@ -0,0 +1,137 @@
+ * \return 0 on success, or various error codes on error.
+ */
+int pcmk_stonith_action(stonith_t *st, const char *target, const char *action,
+                        const char *name, int timeout, int tolerance);
```
Added a patch to convert to unsigned ints and milliseconds.  I haven't decided anything about the stonith_t parameter yet.

clumens replied:
All options I can think of involve abstracting out the stonith_t parameter into some more generic type that could potentially hold multiple kinds of fence things.  And that seems like a larger patch than what I want to add to this PR.  On the other hand, it sure would be nice to have this API start off right without having to do a bunch of other things to it later on.

clumens commented:
Okay, fixed the coverity problems.  All that I have left to do on this is audit and adapt to new return values, and then double check everything.  Then I'll be happy to call this good enough and start looking at converting something else to an API.

kgaillot commented on include/pacemaker.h:
```diff
@@ -0,0 +1,193 @@
+ * \return 0 on success, or various error codes on error.
+ */
+int pcmk_fence_action(stonith_t *st, const char *target, const char *action,
+                      const char *name, int timeout, int tolerance);
```
I've been thinking about the use of stonith_t and stonith_key_value_t in the new API. I think overhauling the stonith IPC API first would be unreasonable at this point, so there's not much of a way around it. However I'm thinking we can #ifdef out the public API here for now, and not release it until we get around to the IPC API. I don't think that will be a problem since this API is mainly an initial demonstration rather than an urgent need.

clumens replied:
Yeah, that's reasonable.  Once everything else is looking good, I'll add a standalone patch that adds the ifdefs.  Hopefully later we can just git revert that single patch.

kgaillot commented:
Looks good to me, let me know when you're ready to merge

clumens commented:
I can't think of anything else that needs to be done at the moment.  Probably the next step is to work on crm_mon, and that might turn up something.  That'll be much more complicated.

kgaillot commented:
Something went wrong here. The fencer is using a timeout of 0 for all fencing operations initiated by stonith_admin. Try running the regression tests (cts-fencing).

clumens commented:
I bet I didn't adjust the timeout values tools/stonith_admin.c is passing.  In other words, it's passing 120 thinking it means seconds, and the library is receiving 120 thinking it means milliseconds.  Then when that gets converted back into seconds before being passed to a stonith_t command, it's ending up as 120/1000 which means no timeout.
